app-api/src/main/java/org/openmrs/eip/app/management/repository/DebeziumOffsetRepository.java

```diff
@@ -0,0 +1,23 @@
+package org.openmrs.eip.app.management.repository;
+
+import java.time.LocalDateTime;
+
+import org.openmrs.eip.app.management.entity.DebeziumOffset;
+import org.springframework.data.jpa.repository.JpaRepository;
+import org.springframework.data.jpa.repository.Modifying;
+import org.springframework.data.jpa.repository.Query;
+import org.springframework.data.repository.query.Param;
+import org.springframework.transaction.annotation.Transactional;
+
+public interface DebeziumOffsetRepository extends JpaRepository<DebeziumOffset, Long> {
```

Comment:
My guess is that you should always have exactly one row in this table, not so?

Reply:
Yes, the first version of this code was like that, as you can see [here](https://github.com/FriendsInGlobalHealth/openmrs-eip/commit/def091be210751ffca97412401aa30577ec3d091#diff-941efea7088456bad90da782c77b336e591ac28d37af948a5ce9f44692bd089bL17).
But, if you recall, during one of our centralization weekly meetings, we said that in the case of a power cut, the offset stored by the debezium ends up pointing to an invalid position in the master (MySQL) due to some information that was not flushed by the DBMS. 

So, currently, when the implementation team faces this issue in live environment, they search for the most recent offset backup pointing to the immediately preceding binlog, and they use this as the current offset file and restart the application. In this way the application can resume the sync without losing any event.

With only one offset in the database it would not be possible to automate this. That’s why we changed the logic to store the offset history, so when we encounter this offset issue, we can go back to the previous valid offset.

Reply:
So how do you determine the valid offset with the valid position